### PR TITLE
FEATURE: Dimension switcher

### PR DESCRIPTION
**Pull request by skurfuerst**
- [x] Get available dimensions and presets from server and render in drop down
- [x] Update allowed presets after selecting a different dimension combination (implemented via guest `metaData` for now, could be improved to use an API service and async update via saga)
- [x] Update selected presets when navigating in the content canvas
- [x] Resolve URI after changing dimensions and update content canvas src

### CODE REVIEW

File: Resources/Private/JavaScript/Host/Containers/SecondaryToolbar/DimensionSwitcher/index.js [hunk 1]
```diff
@@ -1,48 +1,108 @@
...
             <DropDown className={style.dropDown}>
                 <DropDown.Header className={style.dropDown__btn}>
-                    <Icon className={style.dropDown__btnIcon} icon="globe"/>
-                    ToDo: Current dimension name
+                    {contentDimensions.map((dimensionConfiguration, dimensionName) =>
+                        <SelectedPreset
+                            dimensionName={dimensionName}
+                            icon={dimensionConfiguration.get('icon')}
+                            dimensionLabel={dimensionConfiguration.get('label')}
+                            presetLabel={activePresets.get(dimensionName).get('label')} />
```
Comment (skurfuerst) [inline Resources/Private/JavaScript/Host/Containers/SecondaryToolbar/DimensionSwitcher/index.js]: here I'd rather use Plow.js, like `$get([dimensionName, 'label'], activePresets)` -- so to not couple the GET to Immutable.JS.

File: Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    [SELECT_PRESET]: ({name, presetName}) => state => {
+        const dimensionValues = Immutable.fromJS($get(['cr', 'contentDimensions', 'byName', name, 'presets', presetName, 'values'], state));
+        return $set(['cr', 'contentDimensions', 'active', name], dimensionValues, state);
+    }
+});
+
+const active = $get('cr.contentDimensions.active');
+const byName = $get('cr.contentDimensions.byName');
+
+const activePresets = createSelector([
```
Comment (skurfuerst) [inline Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js]: a short docu would be nice, which format the "activePresets" selector returns.

File: Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    active,
+    byName
+], (active, byName) => {
+    return active.map((dimensionValues, name) => {
+        const dimensionConfiguration = byName.get(name);
+        const presets = dimensionConfiguration.get('presets');
+        const activePreset = presets.findKey(preset => preset.get('values').equals(dimensionValues));
+        const presetName = activePreset || dimensionConfiguration.get('defaultPreset');
+        return presets.get(presetName).set('name', presetName);
```
Comment (skurfuerst) [inline Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js]: use Plow again :)

**Comment by skurfuerst** [comment]
Hey @hlubek,

I just checked your code, generally :+1: really nice :)  -just a few nitpicks.

If you like, we can also merge intermediate states, as the feature does not exist at all so far.



All the best,
Sebastian

**Comment by dimaip** [comment]
@skurfuerst Hey, but why are tests failing again?

**Comment by skurfuerst** [comment]
Good question - we need to check this!

**Comment by hlubek** [comment]
@dimaip Agreed, I'll check the test and lint and refactor a bit (including @skurfuerst comments) when the feature works basically.

File: Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js (hunk 3 above)
Comment (hlubek) [inline Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js]: I'd think `$set([presetName, 'name'], presetName, presets)` should work, but it doesn't... any idea?

**Comment by hlubek** [comment]
Yippie, it's working! There are some improvements we could implement:

- [ ] Close dropdown after selecting a preset
- [ ] Refactor passing of allowed preset combinations, because it slows down the rendering of each page load

File: Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+//
+// Export the reducer
+//
+export const reducer = handleActions({
+    [system.INIT]: state => $set(
+        'cr.contentDimensions',
+        new Map({
+            byName: Immutable.fromJS(byName(state)),
+            active: Immutable.fromJS(active(state)),
+            allowedPresets: Immutable.fromJS(allowedPresets(state))
```
Comment (skurfuerst) [inline Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js]: Here, I'd like to have some documentation how the structure in the state actually looks like :-) As that is crucial to understand for the other parts

File: Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        return $set(['cr', 'contentDimensions', 'active', name], dimensionValues, state);
+    },
+    [SET_ACTIVE]: ({dimensionValues}) => state => {
+        const previousActive = active(state);
+        const newActive = previousActive.toSeq().map((values, dimensionName) => new List(dimensionValues[dimensionName])).toMap();
+        return $set('cr.contentDimensions.active', newActive, state);
+    }
+});
+
+const activePresets = createSelector([
```
Comment (skurfuerst) [inline Resources/Private/JavaScript/Host/Redux/CR/ContentDimensions/index.js]: some doumentation how the structure returned by activePresets looks like would be great :)

**Comment by skurfuerst** [comment]
Hey Christopher @hlubek - awesome change :)

Just two nitpicks; once they are solved, feel free to merge the change 👍 

All the best,
Sebastian

**Comment by hlubek** [comment]
Merging with @skurfuerst's permission 😄

**Pull request merged**